### ISSUE TITLE
[api] REST Pagination question

### ISSUE THREAD

**Issue by mateesville93** [code4romania/seismic-risc#538]
We need to add pagination for large querysets.
Now I ask which one is better and friendly with our frontend?
@tudoramariei 

We have three options:

**1. PageNumberPagination:**

**Request:**
`GET http://localhost:8000/api/v1/buildings/?page=4`

**Response:**
```
HTTP 200 OK
{
    "count": 50006
    "next": "http://localhost:8000/api/v1/buildings/?page=5",
    "previous": "http://localhost:8000/api/v1/buildings/?page=3",
    "results": [
       …
    ]
}
```
**2. LimitOffsetPagination:**

**Request:**

`GET http://localhost:8000/api/v1/buildings/?limit=100&offset=400`
**Response:**

```
HTTP 200 OK
{
    "count": 50006
    "next": "http://localhost:8000/api/v1/buildings/?limit=100&offset=500",
    "previous": "http://localhost:8000/api/v1/buildings/?limit=100&offset=300",
    "results": [
       …
    ]
}
```

**3. CursorPagination:**
**Request:**
`GET http://localhost:8000/api/v1/buildings/`

**Response:**
```
HTTP 200 OK
{
    "next": "http://localhost:8000/api/v1/buildings/?cursor=cD0yMDIxLTAyLTA1KzE5JTNBMjIlM0E0MS42NTYwNDIlMkIwMCUzQTAw",
    "previous": null,
    "results": [
       …
    ]
}
```

**Comment by tudoramariei** [comment]
We will have quite a few large query sets in the app and I am wondering what is the best way to handle them but I'm not sure pagination is the way to go for our use-case.
What I believe we need right now is for `GET http://localhost:8000/api/v1/buildings/` to return all the buildings with only the `id`, `lat`, and `lng` variables and `GET http://localhost:8000/api/v1/buildings/<building_id>` to return all the details for a building.
The only way we could limit the number of buildings we need to return is if we query based on the current coordinates of the map. So something like `GET http://localhost:8000/api/v1/buildings/?w_lng=27.4913&e_lng=27.6901&n_lat=47.192&s_lat=127` should cover all the points in the Iasi Metropolitan Area (kind of) with the variables being `west/east longitude` and `north/south latitude` and that would be something we would actually be able to use. 

Pagination would help for navigating **the list** of buildings but wouldn't be of great help trying to display buildings on **a map**.
If we were to implement pagination I'd choose something which has a configurable limit. I think `cursor` and `offset` have a better (or rather a more precise) way of choosing the next item than `page`.

**Comment by JustBeYou** [comment]
For small amounts of data, like a few thousands of buildings with only `id`, `lat`, `lng`, client-side clustering for markers and HTTP caching should be enough for decent performance. (I tested this a few months ago on an application I was working on)
For larger datasets, one approach would be to do the clustering on the server-side. This is pretty similar with what you @tudoramariei said. You can read an interesting article about this here https://geovation.github.io/dynamic-server-side-geo-clustering.

**Comment by myshy93** [comment]
Here Maps has clustering https://developer.here.com/documentation/maps/3.1.22.1/dev_guide/topics/clustering.html

**Comment by myshy93** [comment]
it is still needed ? #567 seems to close this also

**Comment by JustBeYou** [comment]
Not quite, that's client-side clustering, so the entire dataset is still transmitted over the network. 
#573 should reduce the size considerably, but I did not check it yet. It's about 1.5 MiB at the moment with 2500 buildings.

**Comment by tudoramariei** [comment]
Tested it on local and the `/buildings` query is now 182KB and it now takes 170ms and not 2-3 seconds for the query to run.
![Screenshot 2021-03-05 at 12 26 50](https://user-images.githubusercontent.com/2820084/110102980-1250d180-7dae-11eb-8596-89b28624be9b.png)

So while this doesn't exactly answer the whole pagination issue, some improvements have been made to the queries and they're more efficient now.
Also, I tested on the local env because while the staging deployment works, none of the new changes have been deployed for some time and I'm hoping #586 will fix the whole situation.
